src/s390x/zipl.rs

```diff
@@ -0,0 +1,172 @@
+
+    let status = Command::new("zipl")
+        .arg("--verbose")
+        .arg("-p")
+        .arg(zipl_config.path())
+        .arg("-i")
+        .arg(vmlinuz.to_str().unwrap())
+        .arg("-r")
+        .arg(ramfs.to_str().unwrap())
+        .arg("--target")
+        .arg(boot.as_ref().to_str().unwrap())
+        .status()
+        .chain_err(|| "failed to execute zipl")?;
+    check(status).chain_err(|| "installing bootloader")?;
+
+    let status = Command::new("chreipl")
+        .arg(dasd)
+        .status()
+        .chain_err(|| "failed to execute chreipl")?;
+    check(status).chain_err(|| format!("setting {} as boot-device", dasd))?;
+    Ok(())
+}
+
+/// Write boot options. On Z we have to generate config file for zipl
+///
+/// # Arguments
+/// * `boot_config` - path to boot config file
+/// * `zipl_config` - path to zipl config file
+fn generate_zipl_config<P1, P2>(
+    boot_config: P1,
+    zipl_config: &mut P2,
+    firstboot_kargs: &Option<String>,
+) -> Result<()>
+where
+    P1: Sized + AsRef<Path>,
+    P2: Write,
+{
+    eprintln!("Generating zipl config");
+    let firstboot_options = firstboot_kargs.clone().unwrap_or_default();
+    let boot_options = get_boot_kargs(boot_config)?;
+    let s390_options = get_s390x_kargs()?;
+    let config = format!(
+        "{} {} {} ignition.firstboot",
+        boot_options, s390_options, firstboot_options
+    );
+    zipl_config
+        .write_all(config.as_bytes())
+        .chain_err(|| "writing zipl config")?;
+
+    Ok(())
+}
+
+fn find_file<P: AsRef<Path>>(root: P, pat: &str) -> Result<PathBuf> {
+    let mut path = PathBuf::new();
+    path.push(&root);
+    for entry in WalkDir::new(&path).into_iter().filter_map(|e| e.ok()) {
+        if !entry.file_type().is_file() {
+            continue;
+        }
+        let name = entry.file_name().to_string_lossy();
+        if pat.starts_with('.') {
+            if !name.ends_with(pat) {
+                continue;
+            }
+        } else if !name.starts_with(pat) {
+            continue;
+        }
+
+        return Ok(entry.path().to_path_buf());
+    }
+    Err(format!(
+        "Cannot find file with mask: {} in {}",
+        pat,
+        root.as_ref().display()
+    )
+    .into())
+}
+
+fn boot_config<P: AsRef<Path>>(boot: P) -> Result<PathBuf> {
+    let mut path = PathBuf::new();
+    path.push(boot);
+    path.push("loader");
+    path.push("entries");
+    find_file(&path, ".conf")
+}
+
+fn kernel<P: AsRef<Path>>(boot: P) -> Result<PathBuf> {
+    let mut path = PathBuf::new();
+    path.push(boot);
+    path.push("ostree");
+    find_file(&path, "vmlinuz")
+}
+
+fn initramfs<P: AsRef<Path>>(boot: P) -> Result<PathBuf> {
+    let mut path = PathBuf::new();
+    path.push(boot);
+    path.push("ostree");
+    find_file(&path, "initram")
+}
+
+fn get_s390x_kargs() -> Result<String> {
+    let cmdline = read_to_string("/proc/cmdline").chain_err(|| "reading cmdline")?;
```

Comment:
Parsing the kernel cmdline can be subtle, I think https://github.com/coreos/afterburn/ has some Rust code relevant for that?   We also have [C code in libostree](https://github.com/ostreedev/ostree/blob/master/src/libostree/ostree-kernel-args.c).

Reply:
Indeed, but here it's not about parsing, just copying s390x specific `key=value` options